File: 2020/day05/main.go
/**
 * day 05 2020
 * https://adventofcode.com/2020/day/5
 *
 * compile:			go build main.go
 * run:				./main < input
 * compile & run:	go run main.go < input
 **/
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

// get upper 7 bits
func upper7Bits(b int) int {
	return (b >> 3) & 0x7F
}

// get lower 3 bits
func lower3bits(b int) int {
	return b & 0x7
}

// convert the seating chart string to an int
func toBin(s string) int {
	var n int
	for _, c := range s {
		if c == 'B' || c == 'R' {
			n = n << 1 | 1
		} else {
			n = n << 1
		}
	}
	return n
}

func findMyTicket(tkts []int) int {

	sort.Ints(tkts[:])

	// start 5 tickets in
	for i := 5; i < len(tkts); i++ {
		prevTkt := tkts[i] - 1
		nextTkt := tkts[i] + 1

		if tkts[i-1] != prevTkt || tkts[i+1] != nextTkt {
			return tkts[i] + 1
		}
	}
	return 0
}

func main() {
	largestSeat := 0
	var tkts []int // save all tkt numbers to use for part 2

	scan := bufio.NewScanner(os.Stdin)
	for scan.Scan() {
		buf := scan.Text()
		seat := toBin(buf)
		row := upper7Bits(seat)
		col := lower3bits(seat)

		// part 1 - get largest seat number
		seat = row*8 + col
		if seat > largestSeat {
			largestSeat = seat
		}

		tkts = append(tkts, seat)
	}
	mySeat := findMyTicket(tkts)

	fmt.Println("part 1:", largestSeat)
	fmt.Println("part 2:", mySeat)
}

File: 2020/day05/main.py
INPUT = 'input'

with open(INPUT) as f:
	buf = [line.strip() for line in f.readlines()]

largest_seat = 0
tkts = []

def to_bin(ticket):
    n = 0
    for t in ticket:
        # OR in a 1 if char is 'B' or 'R', else a 0
        n = n << 1 | ((t == 'B') or (t == 'R'))
    return n


def find_my_tkt(tkts):
    tkts = sorted(tkts)
    l = len(tkts)
    for i in range(10, l):
        if tkts[i + 1] != tkts[i] + 1:
            # this means there's a missing tkt, which would be mine
            return tkts[i] + 1
    return 0


for ticket in buf:
    seat = to_bin(ticket)       # convert string to binary
    row = (seat >> 3) & 0x7F    # grab the upper 7 bits
    col = seat & 0x7            # grab the lower 3 bits

    # part 1 - get largest seat number
    seat = row * 8 + col
    if seat > largest_seat:
        largest_seat = seat
    tkts.append(seat)

# part 2
my_seat = find_my_tkt(tkts)

assert largest_seat == 838
assert my_seat == 714
print('part 1:', largest_seat)
print('part 2:', my_seat)
